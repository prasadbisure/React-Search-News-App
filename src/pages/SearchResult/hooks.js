import { useState, useEffect, useMemo } from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import { fetchSearchList } from '../../services';

const useSearchHooks = () => {
   
    const [searchResult, setSearchResult] = useState({});
    const [currentPage, setCurrentPage] = useState(1);
    const [pageCount, setPageCount] = useState(0);
    const history = useHistory();
    const ItmesPerPage = 10;

    const  useQuery = () => {
        const { search } = useLocation();
        return useMemo(() => new URLSearchParams(search), [search]);
    }
    let query = useQuery();
    let searchParam;

    const handleOnPageChange = (event) => {
        setCurrentPage(event.selected + 1);
    };

    useEffect(() => {
        
        searchParam = query.get('query');
        fetchSearchList(searchParam, currentPage, ItmesPerPage).then(res => {
            setSearchResult(res?.data?.response?.results);
            setPageCount(res?.data?.response?.pages);
        });
    },[query, currentPage]);

    const handleTagKeywordClick = (param) => {
        history.push({
            pathname: '/search',
            search: `?query=${param}`,            
        })
}

    return {
        searchResult: Array.from(searchResult),
        pageCount,
        isSearchData: Boolean(searchResult.length),
        searchParam,
        handleOnPageChange,
        handleTagKeywordClick 
    }
}
export default useSearchHooks;