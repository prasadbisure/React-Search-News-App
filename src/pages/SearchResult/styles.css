.wrapper {
    width: 60%;
    margin: auto;
    padding: 50px;
}
.thumbnail-wrapper {
    display: flex;
    background: #ccc;
    margin-bottom: 24px;
}

.thumb-img {
    background-color: black;
    width: 250px;
    height: 200px;
    display: inline-block; 
    background-size: cover;
    background-position: center center;
    background-repeat: no-repeat;
}

.thumb-content {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    margin: 1rem;
}
.thumb-content h2 {
    margin: 16px 24px;
}
.thumb-content a {
    text-decoration: unset;
    color: #000;
}

.tags {
    display: flex;
    text-align: right;
    flex-flow: wrap;
    justify-content: end;
}

.tags span {
    margin-right: 1rem;
    margin-top: 5px;
    background: #000;
    color: #fff;
    padding: 2px 8px;
    border-radius: 10px;
    font-size: 11px;
    cursor: pointer;
}

ul.pagination-container {
    list-style: none;
}

ul.pagination-container li {
    display: inline-block;
    background: #545454;
    color: #fff;
    padding: 12px;
    border-radius: 22px;
    margin: 9px;
    cursor: pointer;
}

.no-result {
    background: #ff00008c;
    padding: 6px;
    font-size: 16px;
    font-weight: bold;
}
