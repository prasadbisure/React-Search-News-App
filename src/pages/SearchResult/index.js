import useSearchHooks from "./hooks";
import './styles.css';
import ReactPaginate from 'react-paginate';

const SearchResult = () => {
    const { searchResult, pageCount, isSearchData, handleOnPageChange, handleTagKeywordClick } = useSearchHooks();
    return (
        <div className="wrapper">
            {isSearchData && (searchResult.map(item => (
                <div className="thumbnail-wrapper">
                <div className="thumb-img" style={{ backgroundImage: `url(${item.fields.thumbnail})`}}></div>
                <div className="thumb-content">
                    <a href={item.webUrl} rel="noopener noreferrer" target="_blank"><h2>{item.fields.headline}</h2></a>  
                    <div className="tags">
                        {item.tags.map(tag=>(
                            <span onClick={()=> handleTagKeywordClick(tag.webTitle)}>{tag.webTitle}</span>
                        ))}
                    </div>  
                </div>
            </div>
            )))}
            {isSearchData && (<ReactPaginate 
                breakLabel="..." 
                nextLabel=">" 
                previousLabel="<" 
                onPageChange={handleOnPageChange}
                pageCount={pageCount}
                className="pagination-container"
                pageClassName="pagination-item"
                pageRangeDisplayed={5}
                marginPagesDisplayed={1}
                />)}

            {!isSearchData && (<div className="no-result">
                <p>No result found </p>
                <p><a href="/">Click here</a> to search again.</p>
            </div>)}
        </div>
    );
};

export default SearchResult;
