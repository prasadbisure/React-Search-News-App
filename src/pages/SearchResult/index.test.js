import { shallow } from 'enzyme';
import SearchResult from './';
import  useSearchHooks  from "./hooks";

jest.mock('./hooks', () => jest.fn());

// jest.mock('./hooks', () => {
//     return {
//         __esModule: true,
//         default: () =>  ({
//             searchResult: [{
//                 id: 123,
//                 fields: [],
//                 webUrl: 'http://',
//                 tags: [{
//                     id: 12345,
//                     webTitle:'Web title'
//                 }]
//             }], 
//             pageCount: 100, 
//             isSearchData: false, 
//             handleOnPageChange: jest.fn(), 
//             handleTagKeywordClick: jest.fn()
//         })
//     }
// });


describe('SearchResult component', () => {

    const props = {
        searchResult: [{
            id: 123,
            fields: [],
            webUrl: 'http://',
            tags: [{
                id: 12345,
                webTitle:'Web title'
            }]
        }], 
        pageCount: 100, 
        isSearchData: true, 
        handleOnPageChange: jest.fn(), 
        handleTagKeywordClick: jest.fn()
    }

    
    it('should render snapshots when searchData found', () => {
        useSearchHooks.mockReturnValue({
            ...props
        })
        const wrapper =  shallow(<SearchResult {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });
    
    it('should render snapshots when searchData not found', () => {
        const updatedProps = {
            ...props,
            isSearchData: false
        }
        useSearchHooks.mockReturnValue({
            ...updatedProps
        })
        const wrapper =  shallow(<SearchResult {...updatedProps}/>);
        expect(wrapper).toMatchSnapshot();
    });

})

