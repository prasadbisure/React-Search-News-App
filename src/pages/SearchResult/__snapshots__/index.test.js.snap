// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult component should render snapshots when searchData found 1`] = `
<div
  className="wrapper"
>
  <div
    className="thumbnail-wrapper"
    key="123"
  >
    <div
      className="thumb-img"
      key="123"
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    />
    <div
      className="thumb-content"
      key="123"
    >
      <a
        href="http://"
        key="123"
        rel="noopener noreferrer"
        target="_blank"
      >
        <h2 />
      </a>
      <div
        className="tags"
        key="123"
      >
        <span
          key="123"
        >
          Web title
        </span>
      </div>
    </div>
  </div>
  <PaginationBoxView
    activeClassName="selected"
    breakLabel="..."
    className="pagination-container"
    disableInitialCallback={false}
    disabledClassName="disabled"
    eventListener="onClick"
    hrefAllControls={false}
    marginPagesDisplayed={1}
    nextAriaLabel="Next page"
    nextClassName="next"
    nextLabel=">"
    nextPageRel="next"
    nextRel="next"
    onPageChange={[MockFunction]}
    pageClassName="pagination-item"
    pageCount={100}
    pageLabelBuilder={[Function]}
    pageRangeDisplayed={5}
    prevPageRel="prev"
    prevRel="prev"
    previousAriaLabel="Previous page"
    previousClassName="previous"
    previousLabel="<"
    selectedPageRel="canonical"
  />
</div>
`;

exports[`SearchResult component should render snapshots when searchData not found 1`] = `
<div
  className="wrapper"
>
  <div
    className="no-result"
  >
    <p>
      No result found 
    </p>
    <p>
      <a
        href="/"
      >
        Click here
      </a>
       to search again.
    </p>
  </div>
</div>
`;
